tools/pb2py

```diff
@@ -2,239 +2,213 @@
-                   (number, fieldname, type, flags, comment))
 
-        args.append("        %s: %s = %s," % (fieldname, pytype, value))
+def create_message_import(name):
+    return "from .%s import %s" % (name, name)
 
-        if repeated:
-            assigns.append("        self.%s = [] if %s is None else %s" % (fieldname, fieldname, fieldname))
-        else:
-            assigns.append("        self.%s = %s" % (fieldname, fieldname))
 
-        # print fieldname, number, type, repeated, comment
-        # print v.__dict__
-        # print v.CPPTYPE_STRING
-        # print v.LABEL_REPEATED
-        # print v.enum_type
-        # v.has_default_value, v.default_value
-        # v.label == 3 # repeated
-        # print v.number
+def remove_from_start(s, prefix):
+    if s.startswith(prefix):
+        return s[len(prefix):]
+    else:
+        return s
 
-    if cls.DESCRIPTOR.fields_by_name:
-        out.append("    }")
 
-    if msg_id is not None:
-        out.append("    MESSAGE_WIRE_TYPE = %d" % msg_id)
-        if indexfile is not None:
-            if is_upy:
-                indexfile.write("%s = const(%d)\n" % (t, msg_id))
-            else:
-                indexfile.write("%s = %d\n" % (t, msg_id))
+def process_message_imports(descriptor):
+    imports = set()
+
+    for field in descriptor.fields:
+        if field.type == field.TYPE_MESSAGE:
+            imports.add(field.message_type.name)
+
+    for name in sorted(imports):
+        yield create_message_import(name)
+
+
+def create_init_method(fields):
+    yield         "    def __init__("
+    yield         "        self,"
+    for field in fields:
+        yield     "        %s: %s = None," % (field.name, field.py_type)
+    yield         "        **kwargs"
+    yield         "    ) -> None:"
+    for field in fields:
+        if field.repeated:
+            yield "        self.{0} = {0} if {0} is not None else []".format(field.name)
+        else:
+            yield "        self.{0} = {0}".format(field.name)
+    yield         "        super().__init__(**kwargs)"
+
 
-    # Remove duplicate imports
-    imports = sorted(list(set(imports)))
+def process_message(descriptor, protobuf_module, msg_id, is_upy):
+    logging.debug("Processing message %s", descriptor.name)
 
     if is_upy:
-        imports = ['import protobuf as p'] + imports
+        yield "import protobuf as p"
     else:
-        imports = ['from __future__ import absolute_import',
-                   'from .. import protobuf as p'] + imports
+        yield "from .. import protobuf as p"
+
+    fields = list(parse_field(number, field)
+                  for number, field
+                  in descriptor.fields_by_number.items())
+
+    if any(field.repeated for field in fields):
+        yield "if __debug__:"
+        yield "    try:"
+        yield "        from typing import List"
+        yield "    except ImportError:"
+        yield "        List = None"
```

Comment:
an interesting side-effect of this particular choice (`List = None` as opposed to e.g. `List = {}`) is that it caught CPython redhanded at processing type hints but not having a `typing` module.

that might turn out to be a problem in some circumstances, but maybe it's better to have it?